### DISCUSSION TITLE
ESP32 & Berry & ModBus & save response data into variable

### DISCUSSION THREAD

**Discussion by mrRobot62** [arendst/Tasmota#17149] [Q&A] [answered]
Hi folks,
I'm using an EPEVER XTRA 4415N Solar Charger with RS485 interface

I wrote a Berry-Script which send via ModBusSend xxxxxx commands to the charger to read out the registers . That works fine.
Charger answer and I got a ModbusReceive answer

```
20:14:54.152 MQT: SOLAR_ESP32/solar/RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":4,"StartAddress":13057,"Length":7,"Count":1,"Values":[1999]}}
```
looks good.

**Now my question:**
How can I save this result into a variable inside Berry and work on this Json-Struct?

Currently I send this response directly via MQTT to ioBroker and inside ioBroker I disassemble this JSON-Payload.

Any suggestions?

Thanks in advance

**Comment by barbudor** [comment]
You can add a Berry rule on `Modbusreceived` to get the json as a map object
That will not prevent the MQTT message to still be published

More on Berry rules: https://tasmota.github.io/docs/Berry/#rules

**Comment by mrRobot62** [comment, marked as answer]
uhhh - really ? - that's cool - will check this tomorrow

I can't wait ;-)

`
def rule_one(value, trigger)
   print(value)
end
tasmota.add_rule("ModBusReceiver",rule_one)
`
wrote this without success - mhm nothing happens - I'm a tasmota & berry rookie - sorry 


**Update**
YEAH - works - typo in RuleName 
<img width="825" alt="image" src="https://user-images.githubusercontent.com/949032/202922990-57036330-9303-40b9-8ca7-ebbf4289aee1.png">

@barbudor thanks for your help !

**Reply by UBWH** [reply]
@mrRobot62 :

Tasmota <--> Epever is also a project I am looking at.

Above - you said "I wrote a Berry-Script which send via ModBusSend xxxxxx commands to the charger to read out the registers . "

Is that code public? Or would you be happy to share?

Rob

**Reply by mrRobot62** [reply]
Hi - it's a long time that I worked with berry lang, ESP32 and Tasmota. I had problems with ESP32 (every 2-3 days system crashed and I didn't find a solution, so I decided to change my system.

However , as attachment that I implemented - hopefully you find some solutions for your ideas.

[tasmota32_berry_epever.zip](https://github.com/arendst/Tasmota/files/13394365/tasmota32_berry_epever.zip)

Please note: this was only a test approach not a final program

**Reply by UBWH** [reply]
Thanks.

I am looking at both options:

- Berry
- Scripting (has built in support for modbus send/receive+parse)

I will look out for the crashing every few days - but did see that a memory leak was fixed in some tasmota modbus code recently (?)

**Reply by jetdream** [reply]
@UBWH Do you have any success with Berry or Scripting?
I'm also looking for the solution to poll Modbus RTU inputs board and publish the input states via MQTT.
Berry approach looks straightforward and preferred, but I'm worried about crashes mentioned by @mrRobot62.

**Reply by UBWH** [reply]
I got both BERRY & SCRIPTING to work.

**BERRY:**
There was one drawback for me. See: https://github.com/arendst/Tasmota/discussions/20501
@sfromis did suggest a way to fix this by editing the Tasmota code ... but I want to stay away from using a custom Tasmota build.

Crashes? Did not run this project long enough to know if this was an issue.

**SCRIPT:**
Went back to this method as it is reliable and does not hit MQTT with so much unnecessary traffic.
Mind you - there was a crashing problem (Memory leak) in the SCRIPT method, but I reported that and it has been fixed :-)
https://github.com/arendst/Tasmota/discussions/20430

**Reply by sfromis** [reply]
What I meant was not really you editing the code, more like a more general hint about a plausible enhancement to Tasmota, useful to other users.

**Comment by belveder79** [comment]
Sorry to warm this up again, but I wanted to mention that the Rules do NOT work when requesting HEX data (works for RAW though). It took me quite a while to figure that out...

So this one works
`ModBusSend { "deviceaddress": 0x01, "functioncode": 3, "startaddress": 0x0000, "type": "uint16", "count": 1 }`
but this one does not:
`ModBusSend { "deviceaddress": 0x01, "functioncode": 3, "startaddress": 0x0000, "type": "hex", "count": 1 }`

**Reply by sfromis** [reply]
What do you mean by it not working?

Is it about trying to process hex data with "classic" Tasmota rules? That cannot work seamlessly, as there is a deeper issue about Tasmota rules not supporting hex data, and using hex format like `0x01` or `0x0000` is technically **invalid** JSON. Valid JSON only has decimal numbers or strings, in addition to a few special values like true/false.

While the error of hex in JSON is partially tolerated by Tasmota, in rules it will not parse invalid responses having unquoted hex numbers (or valid quoted hex numbers) back into numbers, but simply leave them as strings, which cannot be processed as if they were numbers.

When I said that invalid hex numbers in JSON are "partially tolerated" by Tasmota, that was because this does not apply to JSON parsing in Berry, where such *invalid* JSON will result in the payload being ignored, not available to processing.

Ideally, I'd say that Tasmota should *always* use fully valid JSON, instead of relying on the leniency of *partially* being able to process hex numbers, but only some of the time.

What you have should be visible in the `ModbusReceived` payload, to clarify what the issue is.

**Reply by belveder79** [reply]
The ModbusSend takes the hex values and converts them to int, so that is all fine (a feature, not a bug :-) ) - however, if you try to query with the "type":"hex" value, you don't get the callback into Berry

Say you add a rule like the one above:

```
def rule_one(value, trigger) print(value) end tasmota.add_rule("ModBusReceived",rule_one)
```

and you issue commands:
```
22:37:39.041 CMD: ModBusSend { "deviceaddress": 1, "functioncode": 3, "startaddress": 0, "type": "uint16", "count": 1 }
22:37:39.053 MQT: stat/windspeeddirection/RESULT = {"ModbusSend":"Done"}
22:37:39.097 MQT: tele/windspeeddirection/RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":0,"Length":7,"Count":1,"Values":[4]}}
22:37:39.115 {'FunctionCode': 3, 'Length': 7, 'Count': 1, 'DeviceAddress': 1, 'Values': [4], 'StartAddress': 0}
22:37:55.794 CMD: ModBusSend { "deviceaddress": 1, "functioncode": 3, "startaddress": 0, "type": "hex", "count": 1 }
22:37:55.811 MQT: stat/windspeeddirection/RESULT = {"ModbusSend":"Done"}
22:37:55.854 MQT: tele/windspeeddirection/RESULT = {"ModbusReceived":{"HEX":[0x01,0x03,0x02,0x00,0x04,0xB9,0x87]}}
```

but in Berry you only get this:

![image](https://github.com/arendst/Tasmota/assets/112880/ac12b154-e50f-439a-ae5c-94e1d5394f4b)

"raw" works also, but "hex" does not.

**Reply by sfromis** [reply]
Yeah, that is exactly what I described, hex numbers are technically **invalid JSON**, but still produced and *partially* accepted by Tasmota.

**Reply by belveder79** [reply]
Yes, I don't think that there is anything wrong with that... it is just not really obvious either... it is sent over mqtt, but not forwarded to a rule...

**Comment by Metaworld82** [comment]
Hey Guys, 

I have a question on save the response data  which are come back from  Tasmota. 
First i send this request on Modbus

23:39:44.455` CMD: ModBusSend {"deviceaddress": 1, "functioncode": 3, "startaddress": 0, "type":"uint16", "count":1}

and get the following answer in the consol

23:39:44.461 MQT: stat/tasmota_685150/RESULT = {"ModbusSend":"Done"}
23:39:44.574 MQT: tele/tasmota_685150/RESULT ={"ModbusReceived":"DeviceAddress":1,"FunctionCode":3,"StartAddress":0,"Length":7,"Count":1,"Values":[275]}}

no i am trying to store the data of the answer with this simple rule:

rule1 on ModbusReceived#Values DO var1 %value% ENDON 

but var1 stayes at zero??? if i choose an other like StartAddress or Length oder Count i get always the right Value saved in Var1.
Is it about the [] after Values:?

Maybe someone can give me a hint.

Kind Regards, Stefan

PS: Sorry i am an absolute tasmota noob and my english is not the best :-)

**Reply by Metaworld82** [reply]
Solution found, the data i want are in an array so the right Syntax is:
on ModbusRecive#Values[1] DO var1 %value% ENDON to get the data from the first register

Kind Regards, Stefan
